File: src/DelegateDecompiler.Tests/DecimalNullTests.cs
using System;
using System.Linq.Expressions;
using NUnit.Framework;

namespace DelegateDecompiler.Tests
{
    [TestFixture]
    public class DecimalNullTests : DecompilerTestsBase
    {
        [Test]
        public void ExpressionWithNullable()
        {
            Expression<Func<decimal, decimal?>> expected = x => (decimal?)x;
            Func<decimal, decimal?> compiled = x => (decimal?)x;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableEqual()
        {
            Expression<Func<decimal?, decimal?, bool>> expected = (x, y) => x == y;
            Func<decimal?, decimal?, bool> compiled = (x, y) => x == y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableNotEqual()
        {
            Expression<Func<decimal?, decimal?, bool>> expected = (x, y) => x != y;
            Func<decimal?, decimal?, bool> compiled = (x, y) => x != y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableGreaterThan()
        {
            Expression<Func<decimal?, decimal?, bool>> expected = (x, y) => x > y;
            Func<decimal?, decimal?, bool> compiled = (x, y) => x > y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableGreaterThanOrEqual()
        {
            Expression<Func<decimal?, decimal?, bool>> expected = (x, y) => x >= y;
            Func<decimal?, decimal?, bool> compiled = (x, y) => x >= y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableLessThan()
        {
            Expression<Func<decimal?, decimal?, bool>> expected = (x, y) => x < y;
            Func<decimal?, decimal?, bool> compiled = (x, y) => x < y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableLessThanOrEqual()
        {
            Expression<Func<decimal?, decimal?, bool>> expected = (x, y) => x <= y;
            Func<decimal?, decimal?, bool> compiled = (x, y) => x <= y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableMul()
        {
            Expression<Func<decimal?, decimal?, decimal?>> expected = (x, y) => x * y;
            Func<decimal?, decimal?, decimal?> compiled = (x, y) => x * y;
            Test(expected, compiled);
        }
        
        [Test]
        public void ExpressionWithNullablePlus()
        {
            Expression<Func<decimal?, decimal?, decimal?>> expected = (x, y) => x + y;
            Func<decimal?, decimal?, decimal?> compiled = (x, y) => x + y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableDiv()
        {
            Expression<Func<decimal?, decimal?, decimal?>> expected = (x, y) => x / y;
            Func<decimal?, decimal?, decimal?> compiled = (x, y) => x / y;
            Test(expected, compiled);
        }
        
        [Test]
        public void ExpressionWithNullableEqual2()
        {
            Expression<Func<decimal?, decimal, bool>> expected = (x, y) => x == y;
            Func<decimal?, decimal, bool> compiled = (x, y) => x == y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableNotEqual2()
        {
            Expression<Func<decimal?, decimal, bool>> expected = (x, y) => x != y;
            Func<decimal?, decimal, bool> compiled = (x, y) => x != y;

            Func<decimal?, decimal, bool> c = (x, y) => (x.GetValueOrDefault() == y) ? (x.HasValue == false) : true;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableGreaterThan2()
        {
            Expression<Func<decimal?, decimal, bool>> expected = (x, y) => x > y;
            Func<decimal?, decimal, bool> compiled = (x, y) => x > y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableGreaterThanOrEqual2()
        {
            Expression<Func<decimal?, decimal, bool>> expected = (x, y) => x >= y;
            Func<decimal?, decimal, bool> compiled = (x, y) => x >= y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableLessThan2()
        {
            Expression<Func<decimal?, decimal, bool>> expected = (x, y) => x < y;
            Func<decimal?, decimal, bool> compiled = (x, y) => x < y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableLessThanOrEqual2()
        {
            Expression<Func<decimal?, decimal, bool>> expected = (x, y) => x <= y;
            Func<decimal?, decimal, bool> compiled = (x, y) => x <= y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableMul2()
        {
            Expression<Func<decimal?, decimal, decimal?>> expected = (x, y) => x * y;
            Func<decimal?, decimal, decimal?> compiled = (x, y) => x * y;
            Test(expected, compiled);
        }
        
        [Test]
        public void ExpressionWithNullablePlus2()
        {
            Expression<Func<decimal?, decimal, decimal?>> expected = (x, y) => x + y;
            Func<decimal?, decimal, decimal?> compiled = (x, y) => x + y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableDiv2()
        {
            Expression<Func<decimal?, decimal, decimal?>> expected = (x, y) => x / y;
            Func<decimal?, decimal, decimal?> compiled = (x, y) => x / y;
            Test(expected, compiled);
        }
        
        [Test]
        public void ExpressionWithNullableEqual3()
        {
            Expression<Func<decimal, decimal?, bool>> expected = (x, y) => x == y;
            Func<decimal, decimal?, bool> compiled = (x, y) => x == y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableNotEqual3()
        {
            Expression<Func<decimal, decimal?, bool>> expected = (x, y) => x != y;
            Func<decimal, decimal?, bool> compiled = (x, y) => x != y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableGreaterThan3()
        {
            Expression<Func<decimal, decimal?, bool>> expected = (x, y) => x > y;
            Func<decimal, decimal?, bool> compiled = (x, y) => x > y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableGreaterThanOrEqual3()
        {
            Expression<Func<decimal, decimal?, bool>> expected = (x, y) => x >= y;
            Func<decimal, decimal?, bool> compiled = (x, y) => x >= y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableLessThan3()
        {
            Expression<Func<decimal, decimal?, bool>> expected = (x, y) => x < y;
            Func<decimal, decimal?, bool> compiled = (x, y) => x < y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableLessThanOrEqual3()
        {
            Expression<Func<decimal, decimal?, bool>> expected = (x, y) => x <= y;
            Func<decimal, decimal?, bool> compiled = (x, y) => x <= y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableMul3()
        {
            Expression<Func<decimal, decimal?, decimal?>> expected = (x, y) => x * y;
            Func<decimal, decimal?, decimal?> compiled = (x, y) => x * y;
            Test(expected, compiled);
        }
        
        [Test]
        public void ExpressionWithNullablePlus3()
        {
            Expression<Func<decimal, decimal?, decimal?>> expected = (x, y) => x + y;
            Func<decimal, decimal?, decimal?> compiled = (x, y) => x + y;
            Test(expected, compiled);
        }

        [Test]
        public void ExpressionWithNullableDiv3()
        {
            Expression<Func<decimal, decimal?, decimal?>> expected = (x, y) => x / y;
            Func<decimal, decimal?, decimal?> compiled = (x, y) => x / y;
            Test(expected, compiled);
        }
        
        [Test, Ignore("Needs optimization")]
        public void ExpressionWithNullableSum3()
        {
            Expression<Func<decimal?, decimal?, decimal?, decimal?>> expected = (x, y, z) => x + y + z;
            Func<decimal?, decimal?, decimal?, decimal?> compiled = (x, y, z) => x + y + z;
            Test(expected, compiled);
        }
    }
}

File: src/DelegateDecompiler.Tests.VB/DecimalNullTests.vb
Imports System
Imports System.Linq.Expressions
Imports NUnit.Framework


<TestFixture>
Public Class DecimalNullTests
    Inherits DecompilerTestsBase


    <Test>
    Public Sub ExpressionWithNullable()
        Dim expected As Expression(Of Func(Of Decimal, Decimal?)) = Function(x As Decimal) x
        Dim compiled As Func(Of Decimal, Decimal?) = Function(x As Decimal) x
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableEqual()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal?, Boolean?)) = Function(x As Decimal?, y As Decimal?) x = y
        Dim compiled As Func(Of Decimal?, Decimal?, Boolean?) = Function(x As Decimal?, y As Decimal?) x = y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableNotEqual()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal?, Boolean?)) = Function(x As Decimal?, y As Decimal?) x <> y
        Dim compiled As Func(Of Decimal?, Decimal?, Boolean?) = Function(x As Decimal?, y As Decimal?) x <> y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableGreaterThan()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal?, Boolean?)) = Function(x As Decimal?, y As Decimal?) x > y
        Dim compiled As Func(Of Decimal?, Decimal?, Boolean?) = Function(x As Decimal?, y As Decimal?) x > y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableGreaterThanOrEqual()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal?, Boolean?)) = Function(x As Decimal?, y As Decimal?) x >= y
        Dim compiled As Func(Of Decimal?, Decimal?, Boolean?) = Function(x As Decimal?, y As Decimal?) x >= y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableLessThan()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal?, Boolean?)) = Function(x As Decimal?, y As Decimal?) x < y
        Dim compiled As Func(Of Decimal?, Decimal?, Boolean?) = Function(x As Decimal?, y As Decimal?) x < y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableLessThanOrEqual()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal?, Boolean?)) = Function(x As Decimal?, y As Decimal?) x <= y
        Dim compiled As Func(Of Decimal?, Decimal?, Boolean?) = Function(x As Decimal?, y As Decimal?) x <= y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableMul()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal?, Decimal?)) = Function(x As Decimal?, y As Decimal?) x * y
        Dim compiled As Func(Of Decimal?, Decimal?, Decimal?) = Function(x As Decimal?, y As Decimal?) x * y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullablePlus()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal?, Decimal?)) = Function(x As Decimal?, y As Decimal?) x + y
        Dim compiled As Func(Of Decimal?, Decimal?, Decimal?) = Function(x As Decimal?, y As Decimal?) x + y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableMinus()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal?, Decimal?)) = Function(x As Decimal?, y As Decimal?) x - y
        Dim compiled As Func(Of Decimal?, Decimal?, Decimal?) = Function(x As Decimal?, y As Decimal?) x - y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableDiv()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal?, Decimal?)) = Function(x As Decimal?, y As Decimal?) x / y
        Dim compiled As Func(Of Decimal?, Decimal?, Decimal?) = Function(x As Decimal?, y As Decimal?) x / y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableEqual2()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal, Boolean?)) = Function(x As Decimal?, y As Decimal) x = y
        Dim compiled As Func(Of Decimal?, Decimal, Boolean?) = Function(x As Decimal?, y As Decimal) x = y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableNotEqual2()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal, Boolean?)) = Function(x As Decimal?, y As Decimal) x <> y
        Dim compiled As Func(Of Decimal?, Decimal, Boolean?) = Function(x As Decimal?, y As Decimal) x <> y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableGreaterThan2()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal, Boolean?)) = Function(x As Decimal?, y As Decimal) x > y
        Dim compiled As Func(Of Decimal?, Decimal, Boolean?) = Function(x As Decimal?, y As Decimal) x > y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableGreaterThanOrEqual2()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal, Boolean?)) = Function(x As Decimal?, y As Decimal) x >= y
        Dim compiled As Func(Of Decimal?, Decimal, Boolean?) = Function(x As Decimal?, y As Decimal) x >= y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableLessThan2()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal, Boolean?)) = Function(x As Decimal?, y As Decimal) x < y
        Dim compiled As Func(Of Decimal?, Decimal, Boolean?) = Function(x As Decimal?, y As Decimal) x < y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableLessThanOrEqual2()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal, Boolean?)) = Function(x As Decimal?, y As Decimal) x <= y
        Dim compiled As Func(Of Decimal?, Decimal, Boolean?) = Function(x As Decimal?, y As Decimal) x <= y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableMul2()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal, Decimal?)) = Function(x As Decimal?, y As Decimal) x * y
        Dim compiled As Func(Of Decimal?, Decimal, Decimal?) = Function(x As Decimal?, y As Decimal) x * y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullablePlus2()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal, Decimal?)) = Function(x As Decimal?, y As Decimal) x + y
        Dim compiled As Func(Of Decimal?, Decimal, Decimal?) = Function(x As Decimal?, y As Decimal) x + y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableMinus2()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal, Decimal?)) = Function(x As Decimal?, y As Decimal) x - y
        Dim compiled As Func(Of Decimal?, Decimal, Decimal?) = Function(x As Decimal?, y As Decimal) x - y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableDiv2()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal, Decimal?)) = Function(x As Decimal?, y As Decimal) x / y
        Dim compiled As Func(Of Decimal?, Decimal, Decimal?) = Function(x As Decimal?, y As Decimal) x / y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableEqual3()
        Dim expected As Expression(Of Func(Of Decimal, Decimal?, Boolean?)) = Function(x As Decimal, y As Decimal?) x = y
        Dim compiled As Func(Of Decimal, Decimal?, Boolean?) = Function(x As Decimal, y As Decimal?) x = y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableNotEqual3()
        Dim expected As Expression(Of Func(Of Decimal, Decimal?, Boolean?)) = Function(x As Decimal, y As Decimal?) x <> y
        Dim compiled As Func(Of Decimal, Decimal?, Boolean?) = Function(x As Decimal, y As Decimal?) x <> y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableGreaterThan3()
        Dim expected As Expression(Of Func(Of Decimal, Decimal?, Boolean?)) = Function(x As Decimal, y As Decimal?) x > y
        Dim compiled As Func(Of Decimal, Decimal?, Boolean?) = Function(x As Decimal, y As Decimal?) x > y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableGreaterThanOrEqual3()
        Dim expected As Expression(Of Func(Of Decimal, Decimal?, Boolean?)) = Function(x As Decimal, y As Decimal?) x >= y
        Dim compiled As Func(Of Decimal, Decimal?, Boolean?) = Function(x As Decimal, y As Decimal?) x >= y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableLessThan3()
        Dim expected As Expression(Of Func(Of Decimal, Decimal?, Boolean?)) = Function(x As Decimal, y As Decimal?) x < y
        Dim compiled As Func(Of Decimal, Decimal?, Boolean?) = Function(x As Decimal, y As Decimal?) x < y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableLessThanOrEqual3()
        Dim expected As Expression(Of Func(Of Decimal, Decimal?, Boolean?)) = Function(x As Decimal, y As Decimal?) x <= y
        Dim compiled As Func(Of Decimal, Decimal?, Boolean?) = Function(x As Decimal, y As Decimal?) x <= y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableMul3()
        Dim expected As Expression(Of Func(Of Decimal, Decimal?, Decimal?)) = Function(x As Decimal, y As Decimal?) x * y
        Dim compiled As Func(Of Decimal, Decimal?, Decimal?) = Function(x As Decimal, y As Decimal?) x * y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullablePlus3()
        Dim expected As Expression(Of Func(Of Decimal, Decimal?, Decimal?)) = Function(x As Decimal, y As Decimal?) x + y
        Dim compiled As Func(Of Decimal, Decimal?, Decimal?) = Function(x As Decimal, y As Decimal?) x + y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableMinus3()
        Dim expected As Expression(Of Func(Of Decimal, Decimal?, Decimal?)) = Function(x As Decimal, y As Decimal?) x - y
        Dim compiled As Func(Of Decimal, Decimal?, Decimal?) = Function(x As Decimal, y As Decimal?) x - y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableDiv3()
        Dim expected As Expression(Of Func(Of Decimal, Decimal?, Decimal?)) = Function(x As Decimal, y As Decimal?) x / y
        Dim compiled As Func(Of Decimal, Decimal?, Decimal?) = Function(x As Decimal, y As Decimal?) x / y
        Test(compiled, expected)
    End Sub

    <Test>
    Public Sub ExpressionWithNullableSum3()
        Dim expected As Expression(Of Func(Of Decimal?, Decimal?, Decimal?, Decimal?)) = Function(x As Decimal?, y As Decimal?, z As Decimal?) x + y + z
        Dim compiled As Func(Of Decimal?, Decimal?, Decimal?, Decimal?) = Function(x As Decimal?, y As Decimal?, z As Decimal?) x + y + z
        Test(compiled, expected)
    End Sub
End Class